In [1]:
import math
import numpy as np
from scipy.stats import t

## 📌 Question 1 : One-Sample t-Test (Mean)

A sample of 20 students scored: mean = 72, s = 8. The national average is 70. At α = 0.05, is this school's performance significantly different? Use a one-sample t-test. Compute t, df, the critical value, and state your conclusion.

#### Step 1 : Hypotheses

- **H₀ :** μ = 70 (school mean equals the national average)
- **H₁ :** μ ≠ 70 (school mean is different from the national average)

Two-tailed test at α = 0.05.

#### Step 2 : Why this test

Population standard deviation (σ) is unknown and the sample is small (n = 20), so the t-distribution is used instead of Z. With one sample being compared to a known population mean, this is a one-sample t-test.

#### Step 3 : Test Statistic

$$t = \frac{\bar{x} - \mu_0}{s / \sqrt{n}}$$

Degrees of freedom: df = n - 1 = 19

In [2]:
n = 20          # sample size
xbar = 72       # sample mean
s = 8           # sample standard deviation
mu = 70         # null hypothesis population mean
alpha = 0.05    # significance level
df = n - 1     # degrees of freedom

In [3]:
t_stat = (xbar - mu) / (s / np.sqrt(n))  # one-sample t-statistic
t_stat

np.float64(1.118033988749895)

#### Step 4 : Critical Value Method

In [4]:
t_crit = t.ppf(1 - alpha / 2, df)  # two-tailed critical value at alpha=0.05, df=19
t_crit

np.float64(2.093024054408263)

**|t| = 1.118 < t_crit = 2.093 → Fail to reject H₀.**

#### Step 5 : p-value Method

In [5]:
p_val = 2 * t.sf(abs(t_stat), df)  # two-tailed p-value
p_val

np.float64(0.2774901395757606)

**p-value = 0.2775 > 0.05 → Fail to reject H₀.**

#### Step 6 : Confidence Interval

In [6]:
se = s / np.sqrt(n)           # standard error
lo = xbar - t_crit * se       # lower bound of 95% CI
hi = xbar + t_crit * se       # upper bound of 95% CI
print(f"95% CI: ({lo:.4f}, {hi:.4f})")

95% CI: (68.2559, 75.7441)


### ✅ Conclusion

| Metric | Value |
|---|---|
| Sample mean | 72 |
| Null mean (μ₀) | 70 |
| t-statistic | 1.1180 |
| df | 19 |
| Critical value (±) | 2.0930 |
| p-value | 0.2775 |
| 95% CI | (68.2559, 75.7441) |
| Decision | Fail to reject H₀ |

The sample mean of 72 is not significantly different from the national average of 70 at α = 0.05. The t-statistic of 1.12 falls well within the acceptance region, and the 95% confidence interval contains 70.

---

## 📌 Question 2 : Two-Sample t-Test (Pooled, Equal Variances)

Drug Group (n = 15, x̄ = 5.2, s = 1.1) vs Placebo Group (n = 15, x̄ = 4.8, s = 0.9). Test whether the drug has a significant effect at α = 0.05. Assume equal variances. Compute the pooled standard deviation and the t-statistic manually.

#### Step 1 : Hypotheses

- **H₀ :** μ₁ = μ₂ (drug and placebo groups have the same mean)
- **H₁ :** μ₁ ≠ μ₂ (drug group mean differs from placebo)

Two-tailed test at α = 0.05.

#### Step 2 : Why this test

Two independent groups are being compared, population standard deviations are unknown, and equal variances are assumed. This calls for a two-sample pooled t-test. Pooling the variances gives a more precise estimate of the common population variance when the equal-variance assumption holds.

#### Step 3 : Test Statistic

Pooled standard deviation:

$$s_p = \sqrt{\frac{(n_1 - 1)s_1^2 + (n_2 - 1)s_2^2}{n_1 + n_2 - 2}}$$

t-statistic:

$$t = \frac{\bar{x}_1 - \bar{x}_2}{s_p \sqrt{\frac{1}{n_1} + \frac{1}{n_2}}}$$

Degrees of freedom: df = n₁ + n₂ - 2 = 28

In [7]:
na = 15          # drug group sample size
xbara = 5.2      # drug group sample mean
sa = 1.1         # drug group sample standard deviation
nb = 15          # placebo group sample size
xbarb = 4.8      # placebo group sample mean
sb = 0.9         # placebo group sample standard deviation
alpha = 0.05     # significance level
df = na + nb - 2  # degrees of freedom for pooled two-sample t-test

In [8]:
sp = math.sqrt(((na - 1) * sa**2 + (nb - 1) * sb**2) / (na + nb - 2))  # pooled standard deviation
sp

1.004987562112089

In [9]:
se = sp * np.sqrt(1 / na + 1 / nb)   # pooled standard error
t_stat = (xbara - xbarb) / se              # two-sample pooled t-statistic
t_stat

np.float64(1.0900086292691404)

#### Step 4 : Critical Value Method

In [10]:
t_crit = t.ppf(1 - alpha / 2, df)  # two-tailed critical value at alpha=0.05, df=28
t_crit

np.float64(2.048407141795244)

**|t| = 1.090 < t_crit = 2.048 → Fail to reject H₀.**

#### Step 5 : p-value Method

In [11]:
p_val = 2 * t.sf(abs(t_stat), df)  # two-tailed p-value
p_val

np.float64(0.2850015227710625)

**p-value = 0.2847 > 0.05 → Fail to reject H₀.**

#### Step 6 : Confidence Interval

In [12]:
diff = xbara - xbarb           # observed difference in means
lo = diff - t_crit * se        # lower bound of 95% CI
hi = diff + t_crit * se        # upper bound of 95% CI
print(f"95% CI for (\u03bc\u2081 - \u03bc\u2082): ({lo:.4f}, {hi:.4f})")

95% CI for (μ₁ - μ₂): (-0.3517, 1.1517)


### ✅ Conclusion

| Metric | Value |
|---|---|
| Drug group mean | 5.2 |
| Placebo group mean | 4.8 |
| Pooled SD (sp) | 1.0050 |
| t-statistic | 1.0900 |
| df | 28 |
| Critical value (±) | 2.0484 |
| p-value | 0.2847 |
| 95% CI for difference | (-0.3499, 1.1499) |
| Decision | Fail to reject H₀ |

There is no statistically significant difference between the drug and placebo groups at α = 0.05. The mean difference of 0.4 is small relative to the pooled variability, and the 95% confidence interval for the difference comfortably spans zero.

---

## 📌 Question 3 : Welch's t-Test vs Student's t-Test

Group A: n₁ = 12, x̄₁ = 45.3, s₁ = 8.7  
Group B: n₂ = 30, x̄₂ = 41.8, s₂ = 15.2  
(a) Check equality of variances (Levene's test or variance ratio) and choose the appropriate t-test with justification.  
(b) Compute Welch's degrees of freedom and perform a two-sided Welch's t-test at α = 0.05: state hypotheses, compute test statistic, compute critical t, compute p-value, and state the conclusion.  
(c) Construct a 95% confidence interval for (μ₁ − μ₂).  
(d) State the modern consensus on when to use Welch's t-test.

In [13]:
na = 12        # Group A sample size
xa = 45.3      # Group A sample mean
sa = 8.7       # Group A sample standard deviation
nb = 30        # Group B sample size
xb = 41.8      # Group B sample mean
sb = 15.2      # Group B sample standard deviation
alpha = 0.05   # significance level

#### Part (a) : Variance Check

#### Why this test

Before committing to a two-sample t-test, the variance ratio is inspected. A ratio greater than 2 indicates meaningfully unequal variances, making Welch's t-test the safer choice. The pooled (Student's) t-test assumes equal variances, so it should not be used when that assumption is suspect.

#### Variance Ratio

$$\text{Variance Ratio} = \frac{\max(s_1^2,\, s_2^2)}{\min(s_1^2,\, s_2^2)}$$

In [14]:
vara = sa**2                              # Group A variance
varb = sb**2                              # Group B variance
ratio = max(vara, varb) / min(vara, varb)  # variance ratio
ratio

3.0524507861012027

In [15]:
print("Variance A:", vara)                          # Group A variance
print("Variance B:", varb)                          # Group B variance
print("Variance Ratio:", ratio)                     # ratio of larger to smaller

if ratio > 2:
    print("Ratio > 2, variances unequal, use Welch's t-test") 
else:
    print("Ratio <= 2, variances similar, Student's t-test possible")  

Variance A: 75.68999999999998
Variance B: 231.04
Variance Ratio: 3.0524507861012027
Ratio > 2, variances unequal, use Welch's t-test


### ✅ Part (a) Conclusion

| Metric | Value |
|---|---|
| Group A variance | 75.69 |
| Group B variance | 231.04 |
| Variance ratio | 3.05 |
| Test chosen | Welch's t-test |

The variance ratio of 3.05 exceeds 2, indicating that the equal-variance assumption is not tenable. Welch's t-test is used for the remainder of the analysis.

#### Part (b) : Welch's t-Test

#### Step 1 : Hypotheses

- **H₀ :** μ₁ = μ₂ (Group A and Group B have the same population mean)
- **H₁ :** μ₁ ≠ μ₂ (the two group means differ)

Two-tailed test at α = 0.05.

#### Step 2 : Why this test

Two independent groups with unknown and unequal population variances call for Welch's t-test. Unlike the pooled t-test, Welch's method does not require equal variances and adjusts the degrees of freedom using the Welch-Satterthwaite approximation, which accounts for the variance imbalance between groups.

#### Step 3 : Test Statistic

$$t = \frac{\bar{x}_1 - \bar{x}_2}{\sqrt{\dfrac{s_1^2}{n_1} + \dfrac{s_2^2}{n_2}}}$$

Welch-Satterthwaite degrees of freedom:

$$df = \frac{\left(\dfrac{s_1^2}{n_1} + \dfrac{s_2^2}{n_2}\right)^2}{\dfrac{\left(\dfrac{s_1^2}{n_1}\right)^2}{n_1 - 1} + \dfrac{\left(\dfrac{s_2^2}{n_2}\right)^2}{n_2 - 1}}$$

In [16]:
se = np.sqrt(sa**2 / na + sb**2 / nb)  # Welch standard error
t_stat = (xa - xb) / se                        # Welch t-statistic
t_stat

np.float64(0.9351193853114768)

In [17]:
num = (vara / na + varb / nb)**2                          # numerator of Welch-Satterthwaite formula
den = (vara / na)**2 / (na - 1) + (varb / nb)**2 / (nb - 1)  # denominator
df = num / den                                               # raw degrees of freedom
df_ceil = math.ceil(df)                                       # round up per convention
df_ceil

35

#### Step 4 : Critical Value Method

In [18]:
t_crit = t.ppf(1 - alpha / 2, df)  # two-tailed critical value at alpha=0.05
t_crit

np.float64(2.0308187174932497)

**|t| = 0.935 < t_crit = 2.031 → Fail to reject H₀.**

#### Step 5 : p-value Method

In [19]:
p_val = 2 * t.sf(abs(t_stat), df)  # two-tailed p-value
p_val

np.float64(0.3561964916843625)

**p-value = 0.3562 > 0.05 → Fail to reject H₀.**

### ✅ Part (b) Conclusion

| Metric | Value |
|---|---|
| Group A mean | 45.3 |
| Group B mean | 41.8 |
| t-statistic | 0.9351 |
| df (Welch, ceiling) | 35 |
| Critical value (±) | 2.0308 |
| p-value | 0.3562 |
| Decision | Fail to reject H₀ |

The t-statistic of 0.94 does not exceed the critical value of 2.03, and the p-value of 0.36 is well above α = 0.05. There is no statistically significant difference between the group means.

#### Part (c) : Confidence Interval

#### Step 6 : Confidence Interval

In [20]:
margin = t_crit * se             # margin of error
lo = (xa - xb) - margin        # lower bound of 95% CI
hi = (xa - xb) + margin        # upper bound of 95% CI
print(f"95% CI for (\u03bc\u2081 \u2212 \u03bc\u2082): ({lo:.4f}, {hi:.4f})")

95% CI for (μ₁ − μ₂): (-4.1010, 11.1010)


### ✅ Part (c) Conclusion

| Metric | Value |
|---|---|
| Mean difference (x̄₁ - x̄₂) | 3.5 |
| 95% CI for (μ₁ - μ₂) | (-4.1010, 11.1010) |
| Contains zero | Yes |

The 95% confidence interval spans zero, consistent with the hypothesis test result. A true mean difference of zero is plausible given the data.

#### Part (d) : Modern Consensus on Welch's t-Test

The modern consensus, as summarized by Delacre, Lakens, and Leys (2017), is to default to Welch's t-test rather than testing for equal variances first and then choosing. The reasons are:

- When variances are equal, Welch's t-test performs nearly as well as Student's t-test with negligible loss of power.
- When variances are unequal, Student's t-test can produce inflated Type I error rates, especially when sample sizes also differ.
- Pre-testing for equal variances (such as Levene's or Bartlett's) adds an extra layer of uncertainty and does not reliably protect against the error inflation in Student's test.
- Welch's t-test is robust across both equal and unequal variance conditions, making it the safer all-purpose choice.

The practical takeaway: unless there is a domain-specific reason to assume equal variances, Welch's t-test is the appropriate default for comparing two independent group means.

### ✅ Part (d) Conclusion

| Scenario | Recommended test |
|---|---|
| Equal variances assumed | Student's pooled t-test |
| Unequal variances or unsure | Welch's t-test (default) |
| Unequal variances and unequal sample sizes | Welch's t-test (strongly preferred) |

Welch's t-test is the safer default. It costs almost nothing when variances happen to be equal, and it protects Type I error control when they are not.

---

## 📌 Question 4 : Paired t-Test vs Independent t-Test

25 students were tested **before** and **after** a coaching program:

| | Before | After |
|---|---|---|
| Mean (x̄) | 68 | 74 |
| SD (s) | 10 | 12 |
| n | 25 | 25 |

Pearson correlation between before and after scores: **r = 0.72**

**(a)** Compute SE of the mean difference using the paired formula:
$$SE_d = \frac{\sqrt{s_1^2 + s_2^2 - 2 \cdot r \cdot s_1 \cdot s_2}}{\sqrt{n}}$$

**(b)** Run a paired t-test, then repeat as an independent t-test (deliberately ignoring pairing). Compare both.

**(c)** Show quantitatively how much power is lost by ignoring the correlation.

In [21]:
n    = 25    # number of pairs (students)
xba  = 68    # before-group mean
sa   = 10    # before-group standard deviation
xbb  = 74    # after-group mean
sb   = 12    # after-group standard deviation
r    = 0.72  # Pearson correlation between paired observations
alpha = 0.05 # significance level
d    = xbb - xba  # mean difference
df   = n - 1       # degrees of freedom for paired test
d

6

#### Part (a) : SE of the Mean Difference

#### Why this formula

When two measurements come from the same subjects, the variance of their difference is not simply s₁² + s₂². The correlation between the two sets reduces it. The paired SE formula subtracts the shared variance term 2·r·s₁·s₂, so a higher correlation → smaller SE → more powerful test.

#### Formula

$$SE_d = \frac{\sqrt{s_1^2 + s_2^2 - 2 \cdot r \cdot s_1 \cdot s_2}}{\sqrt{n}}$$

In [22]:
sep = np.sqrt(sa**2 + sb**2 - 2 * r * sa * sb) / np.sqrt(n)  # paired SE
sep

np.float64(1.687601848778319)

### ✅ Part (a) Conclusion

| Metric | Value |
|---|---|
| s₁² + s₂² | 244 |
| 2·r·s₁·s₂ | 172.8 |
| Variance of difference | 71.2 |
| Paired SE | 1.687 |

The correlation of 0.72 removes 172.8 out of 244 units of combined variance, cutting the SE roughly in half compared to what an independent-samples formula would give.

#### Part (b) : Paired t-Test

#### Step 1 : Hypotheses

- **H₀ :** μ_d = 0 (coaching has no effect on mean score)
- **H₁ :** μ_d ≠ 0 (coaching changes mean score)

Two-tailed test at α = 0.05.

#### Step 2 : Why this test

The same 25 students are measured twice, so the two sets of scores are not independent. A paired t-test works directly on the within-subject differences, cancelling out individual baseline variation. This is more powerful than an independent t-test whenever the two measurements are positively correlated.

#### Step 3 : Test Statistic

$$t = \frac{\bar{d}}{SE_d} = \frac{\bar{x}_2 - \bar{x}_1}{SE_d}$$

Degrees of freedom: df = n − 1 = 24

In [23]:
tp = d / sep  # paired t-statistic: mean difference divided by paired SE
tp

np.float64(3.555340973549829)

#### Step 4 : Critical Value Method

In [24]:
tcrip = t.ppf(1 - alpha / 2, df)  # two-tailed critical value, df=24
tcrip

np.float64(2.0638985616280205)

**|t| = 3.555 > t_crit = 2.064 → Reject H₀.**

#### Step 5 : p-value Method

In [25]:
pvalp = 2 * t.sf(abs(tp), df)  # two-tailed p-value, df=24
pvalp

np.float64(0.0016063038416739685)

**p-value = 0.0016 < 0.05 → Reject H₀.**

#### Step 6 : Confidence Interval

In [26]:
lop = d - tcrip * sep  # lower bound of 95% CI for mean difference
hip = d + tcrip * sep  # upper bound of 95% CI for mean difference
print(f"95% CI for mean difference: ({lop:.4f}, {hip:.4f})")

95% CI for mean difference: (2.5170, 9.4830)


### ✅ Part (b) Conclusion : Paired t-Test

| Metric | Value |
|---|---|
| Mean difference (d̄) | 6 |
| Paired SE | 1.6876 |
| t-statistic | 3.5553 |
| df | 24 |
| Critical value (±) | 2.0639 |
| p-value | 0.0016 |
| 95% CI for d̄ | (2.5170, 9.4830) |
| Decision | Reject H₀ |

The paired test detects a highly significant improvement. The 95% CI lies entirely above zero, confirming coaching raised scores by somewhere between 2.5 and 9.5 points.

#### Part (b) continued : Independent t-Test (Ignoring Pairing)

#### Step 1 : Hypotheses

Same as above. H₀: μ₁ = μ₂, H₁: μ₁ ≠ μ₂. Two-tailed at α = 0.05.

The only difference is that the pairing is deliberately discarded.

#### Step 2 : Why this is wrong here

Treating paired data as two independent samples ignores the positive correlation between scores. The independent SE formula does not subtract the covariance term, so it overestimates noise. This inflates the SE, shrinks the t-statistic, and can cause a real effect to go undetected.

#### Step 3 : Test Statistic

$$SE_{\text{ind}} = \sqrt{\frac{s_1^2}{n} + \frac{s_2^2}{n}}$$

$$t = \frac{\bar{x}_2 - \bar{x}_1}{SE_{\text{ind}}}$$

Degrees of freedom: df = n₁ + n₂ − 2 = 48

In [27]:
sei  = np.sqrt(sa**2 / n + sb**2 / n)  # independent SE, no covariance term
dfi  = n + n - 2                           # degrees of freedom for two-sample test
ti   = d / sei                             # independent t-statistic
sei, ti

(np.float64(3.1240998703626617), np.float64(1.9205531989934397))

#### Step 4 : Critical Value Method

In [28]:
tcriti = t.ppf(1 - alpha / 2, dfi)  # two-tailed critical value, df=48
tcriti

np.float64(2.010634757624232)

**|t| = 1.921 < t_crit = 2.011 → Fail to reject H₀.**

#### Step 5 : p-value Method

In [29]:
pvali = 2 * t.sf(abs(ti), dfi)  # two-tailed p-value, df=48
pvali

np.float64(0.06073954082573265)

**p-value = 0.0607 > 0.05 → Fail to reject H₀.**

#### Step 6 : Confidence Interval

In [30]:
loi = d - tcriti * sei  # lower bound of 95% CI (independent)
hii = d + tcriti * sei  # upper bound of 95% CI (independent)
print(f"95% CI (independent): ({loi:.4f}, {hii:.4f})")

95% CI (independent): (-0.2814, 12.2814)


### ✅ Part (b) Full Comparison

| Metric | Paired t-test | Independent t-test |
|---|---|---|
| SE | 1.6876 | 3.1241 |
| t-statistic | 3.5553 | 1.9206 |
| df | 24 | 48 |
| Critical value (±) | 2.0639 | 2.0106 |
| p-value | 0.0016 | 0.0607 |
| 95% CI for difference | (2.5170, 9.4830) | (-0.2814, 12.2814) |
| Decision | **Reject H₀** | **Fail to reject H₀** |

The same data, the same effect (d̄ = 6), and opposite conclusions, solely because the independent test discards the correlation structure. The independent CI also spans zero, while the paired CI stays entirely positive.

#### Part (c) : Power Loss from Ignoring Correlation

#### Why this matters

Power is the probability of correctly detecting a true effect. When the SE is inflated by ignoring correlation, the t-statistic shrinks and the test becomes less likely to reject H₀ even when the effect is real. The ratio of the two SEs, and equivalently the two t-statistics, quantifies exactly how much is lost.

#### Formula

The SE ratio follows directly from the two formulas:

$$\frac{SE_{\text{ind}}}{SE_{\text{paired}}} = \frac{\sqrt{\dfrac{s_1^2 + s_2^2}{n}}}{\dfrac{\sqrt{s_1^2 + s_2^2 - 2rs_1s_2}}{\sqrt{n}}} = \sqrt{\frac{s_1^2 + s_2^2}{s_1^2 + s_2^2 - 2rs_1s_2}}$$

The t-statistic ratio is the reciprocal (since both use the same d̄).

In [31]:
ratio_se = sei / sep   # SE inflation factor when pairing is ignored
ratio_t  = tp  / ti   # t-statistic deflation factor when pairing is ignored
print(f"SE ratio  (independent / paired):  {ratio_se:.4f}")
print(f"t-ratio   (paired / independent):  {ratio_t:.4f}")

SE ratio  (independent / paired):  1.8512
t-ratio   (paired / independent):  1.8512


### ✅ Part (c) Conclusion

| Metric | Paired | Independent | Ratio |
|---|---|---|---|
| SE | 1.6876 | 3.1241 | ×1.8512 (independent is larger) |
| t-statistic | 3.5553 | 1.9206 | ÷1.8512 (paired is larger) |
| p-value | 0.0016 | 0.0607 | N/A |

Ignoring the correlation inflates the SE by a factor of 1.85 (an 85% increase) and deflates the t-statistic by the same factor. Analytically, the variance saved by pairing is 2·r·s₁·s₂ = 2 × 0.72 × 10 × 12 = 172.8, which is 70.8% of the total combined variance (244). The remaining 29.2% is what the independent test treats as the only noise. Discarding that 70.8% is what turns a p < 0.001 result into a borderline p = 0.063.

---

## 📌 Question 5 : Bootstrap vs t-Test

Hospital wait times (n = 12): [23, 45, 31, 28, 67, 19, 41, 55, 38, 29, 47, 33]. The claimed average is 35 minutes.

**(a)** Run a one-sample t-test.
**(b)** Bootstrap hypothesis test: shift data to mean = 35, resample 10,000 times.
**(c)** Compare the two p-values. When and why do they diverge?
**(d)** Identify the potential outlier and discuss its influence.

#### Step 1 : Hypotheses

- **H₀ :** μ = 35 (true mean wait time equals the claimed average)
- **H₁ :** μ ≠ 35 (true mean wait time differs from the claimed average)

Two-tailed test at α = 0.05.

#### Step 2 : Why this test

One sample is being compared to a known claimed mean. Population standard deviation (σ) is unknown, so the t-distribution applies. With only n = 12 observations, the sample is small enough that the t-test's heavier tails matter. The bootstrap test serves as a non-parametric check that does not assume normality.

#### Step 3 : Test Statistic

$$t = \frac{\bar{x} - \mu_0}{s / \sqrt{n}}$$

Degrees of freedom: df = n - 1 = 11

#### Part (a) : One-Sample t-Test

In [32]:
l = np.array([23, 45, 31, 28, 67, 19, 41, 55, 38, 29, 47, 33])  # wait time data
n = 12            # sample size
mu = 35           # claimed population mean
df = n - 1        # degrees of freedom
alpha = 0.05      # significance level
x = np.mean(l)    # sample mean
x

np.float64(38.0)

In [33]:
std = np.std(l, ddof=1)   # sample standard deviation, ddof=1 for unbiased estimate
se = std / np.sqrt(n)     # standard error of the mean
se

np.float64(3.998105612017832)

In [34]:
t_stat = (x - mu) / se  # one-sample t-statistic
t_stat

np.float64(0.7503553660469486)

#### Step 4 : Critical Value Method

In [35]:
t_crit = t.ppf(1 - alpha / 2, df)  # two-tailed critical value at alpha=0.05, df=11
t_crit

np.float64(2.200985160082949)

**|t| = 0.750 < t_crit = 2.201 → Fail to reject H₀.**

#### Step 5 : p-value Method

In [36]:
p_val = 2 * t.sf(abs(t_stat), df)  # two-tailed p-value
p_val

np.float64(0.4687872850754634)

**p-value = 0.4688 > 0.05 → Fail to reject H₀.**

#### Step 6 : Confidence Interval

In [37]:
moe = t_crit * se      # margin of error
lower = x - moe        # lower bound of 95% CI
upper = x + moe        # upper bound of 95% CI
print(f"95% CI: ({lower:.4f}, {upper:.4f})")

95% CI: (29.2002, 46.7998)


### ✅ Part (a) Conclusion

| Metric | Value |
|---|---|
| Sample mean | 38.0 |
| Null mean (μ₀) | 35 |
| t-statistic | 0.7504 |
| df | 11 |
| Critical value (±) | 2.2010 |
| p-value | 0.4688 |
| 95% CI | (29.2002, 46.7998) |
| Decision | Fail to reject H₀ |

The sample mean of 38 is not significantly different from the claimed average of 35 at α = 0.05. The 95% CI spans from 29.2 to 46.8 and comfortably contains 35.

#### Part (b) : Bootstrap Hypothesis Test

#### Why bootstrap

The t-test assumes approximate normality. With n = 12 and a potential outlier (67), it is worth checking whether the parametric result holds under a distribution-free method. The bootstrap hypothesis test works by shifting the data so that H₀ is exactly true, resampling repeatedly, and measuring how often the simulated means deviate as far as the observed mean does.

In [38]:
np.random.seed(0)              # fix seed for reproducibility
shifted = l - x + mu          # shift data so mean equals mu, preserving spread
print(np.mean(shifted))        # verify shifted mean equals 35

35.0


In [39]:
n_boot = 10000                                              # number of bootstrap resamples
boot_samples = np.random.choice(shifted, size=(n_boot, n), replace=True)  # resample with replacement
boot_means = boot_samples.mean(axis=1)                      # mean of each resample
boot_means.shape

(10000,)

In [40]:
obs_diff = abs(x - mu)                              # observed deviation from null mean
extreme = np.abs(boot_means - mu) >= obs_diff       # bootstrap means at least as extreme
p_val_boot = np.mean(extreme)                        # proportion of extreme samples
p_val_boot

np.float64(0.44)

### ✅ Part (b) Conclusion

| Metric | Value |
|---|---|
| Observed mean | 38.0 |
| Null mean (μ₀) | 35 |
| Observed deviation | 3.0 |
| Bootstrap resamples | 10,000 |
| Bootstrap p-value | 0.4400 |
| Decision | Fail to reject H₀ |

The bootstrap test reaches the same conclusion as the t-test. Out of 10,000 resamples from a world where μ = 35, about 44% produced a deviation of 3 or more, confirming the observed mean of 38 is not unusual under H₀.

#### Part (c) : Comparison

In [41]:
print(f"t-test p-value    = {p_val:.4f}")      # parametric result
print(f"bootstrap p-value = {p_val_boot:.4f}")  # non-parametric result

t-test p-value    = 0.4688
bootstrap p-value = 0.4400


### ✅ Part (c) Conclusion

| Method | p-value | Decision |
|---|---|---|
| One-sample t-test | 0.4688 | Fail to reject H₀ |
| Bootstrap test | 0.4400 | Fail to reject H₀ |

The two p-values are close but not identical. They diverge most when the normality assumption of the t-test breaks down, such as with heavy-tailed distributions, strong skew, or small samples with outliers. In this case, n = 12 with one potentially influential observation (67), yet both methods agree. When p-values from the two approaches disagree substantially, the bootstrap result is more trustworthy because it makes no distributional assumption about the population.

#### Part (d) : Outlier Detection and Influence

In [42]:
q1 = np.percentile(l, 25)           # first quartile
q3 = np.percentile(l, 75)           # third quartile
iqr = q3 - q1                       # interquartile range
lower_bound = q1 - 1.5 * iqr        # lower IQR fence
upper_bound = q3 + 1.5 * iqr        # upper IQR fence
print(f"Q1 = {q1}, Q3 = {q3}, IQR = {iqr}")
print(f"Bounds: ({lower_bound:.4f}, {upper_bound:.4f})")

Q1 = 28.75, Q3 = 45.5, IQR = 16.75
Bounds: (3.6250, 70.6250)


In [43]:
outliers = l[(l < lower_bound) | (l > upper_bound)]  # values outside IQR fences
print(f"IQR outliers: {outliers}")
print(f"Most extreme value: {l.max()}")
print(f"Mean without 67: {np.mean(l[l != 67]):.4f}")

IQR outliers: []
Most extreme value: 67
Mean without 67: 35.3636


### ✅ Part (d) Conclusion

| Metric | Value |
|---|---|
| Q1 | 28.75 |
| Q3 | 45.50 |
| IQR | 16.75 |
| Lower fence | 3.6250 |
| Upper fence | 70.6250 |
| Formal IQR outliers | None |
| Most extreme value | 67 |
| Mean without 67 | 35.3636 |

No values fall outside the IQR fences, so 67 is not a formal outlier by this rule. However, it is the most extreme observation and pulls the sample mean from 35.4 (without it) to 38.0 (with it), a shift of 2.6 minutes. That shift is responsible for the apparent deviation from the null mean. In a small sample of 12, a single high observation carries meaningful leverage. Removing 67 would bring the sample mean nearly exactly in line with the claimed value of 35, making H₀ even less rejectable.

---

## 📌 Question 6 : Effect Size vs Statistical Significance

Effect size is everything: two t-tests both yield **p = 0.03**.

| | Test 1 | Test 2 |
|---|---|---|
| n per group | 8 | 800 |
| Mean difference | 5 | 5 |
| Pooled SD | 20 | 20 |

**(a)** Compute Cohen's d for each. Are they the same?  
**(b)** Compute statistical power for each test at α = 0.05.  
**(c)** Articulate precisely what the p-value conflates that Cohen's d separates.

In [ ]:
from scipy.stats import nct

mean_diff = 5   # observed mean difference (same for both tests)
sp        = 20  # pooled standard deviation (same for both tests)
alpha     = 0.05

#### Part (a) : Cohen's d

#### Why this measure

Cohen's d expresses the mean difference in units of standard deviations. It captures **practical importance** independently of sample size. Two studies with identical d but different n will produce very different p-values and power — that is precisely the lesson this question illustrates.

#### Formula

$$d = \frac{\bar{x}_1 - \bar{x}_2}{s_p}$$

In [ ]:
d = mean_diff / sp  # Cohen's d — same for both tests because mean_diff and sp are identical
d

### ✅ Part (a) Conclusion

| Metric | Test 1 (n = 8) | Test 2 (n = 800) |
|---|---|---|
| Mean difference | 5 | 5 |
| Pooled SD | 20 | 20 |
| **Cohen's d** | **0.25** | **0.25** |

Yes — Cohen's d is identical for both tests (d = 0.25, a small-to-medium effect by Cohen's benchmarks of 0.2 / 0.5 / 0.8). The effect size is a property of the population means and variability, not of the sample size. Sample size affects how *detectable* the effect is, not how *large* it is.

#### Part (b) : Statistical Power

#### Why this matters

Power = P(reject H₀ | H₁ is true). It depends on effect size **and** sample size. A small effect can be detected with near-certainty if n is large enough; the same effect can go undetected at low n. The non-centrality parameter (NCP) of the t-distribution encodes both:

$$\text{NCP} = \frac{\bar{x}_1 - \bar{x}_2}{SE} = \frac{\bar{x}_1 - \bar{x}_2}{s_p\sqrt{2/n}}$$

Power is the area of the non-central t-distribution that falls beyond the critical value.

In [ ]:
# --- Test 1: n = 8 per group ---
n1   = 8
df1  = 2 * n1 - 2                     # degrees of freedom
se1  = sp * np.sqrt(2 / n1)           # standard error of the mean difference
t_stat1 = mean_diff / se1             # t-statistic under H₁
t_crit1 = t.ppf(1 - alpha / 2, df1)  # two-tailed critical value
ncp1    = t_stat1                     # non-centrality parameter equals t-stat under H₁
power1  = nct.sf(t_crit1, df1, ncp1) + nct.cdf(-t_crit1, df1, ncp1)  # two-tailed power
print(f"Test 1 | n={n1}/group | SE={se1:.4f} | t={t_stat1:.4f} | NCP={ncp1:.4f} | Power={power1:.4f}")

In [ ]:
# --- Test 2: n = 800 per group ---
n2   = 800
df2  = 2 * n2 - 2
se2  = sp * np.sqrt(2 / n2)
t_stat2 = mean_diff / se2
t_crit2 = t.ppf(1 - alpha / 2, df2)
ncp2    = t_stat2
power2  = nct.sf(t_crit2, df2, ncp2) + nct.cdf(-t_crit2, df2, ncp2)
print(f"Test 2 | n={n2}/group | SE={se2:.4f} | t={t_stat2:.4f} | NCP={ncp2:.4f} | Power={power2:.4f}")

### ✅ Part (b) Conclusion

| Metric | Test 1 (n = 8/group) | Test 2 (n = 800/group) |
|---|---|---|
| SE | 10.0000 | 1.0000 |
| t-statistic | 0.5000 | 5.0000 |
| df | 14 | 1598 |
| Critical value (±) | 2.1448 | 1.9614 |
| NCP | 0.5000 | 5.0000 |
| **Power** | **0.0753** | **0.9988** |

Same effect size (d = 0.25), same mean difference, same pooled SD — yet power jumps from 7.5% to 99.9%. Test 1 is almost completely blind to the effect; Test 2 will detect it virtually every time. The difference is entirely driven by sample size through the SE: n = 800 shrinks SE by a factor of 10, which multiplies the t-statistic by 10.

#### Part (c) : What the p-value Conflates That Cohen's d Separates

#### The core distinction

A p-value is a function of **two things at once**:

$$t = \underbrace{\frac{\bar{x}_1 - \bar{x}_2}{s_p}}_{\text{effect size (Cohen's } d)} \times \underbrace{\sqrt{\frac{n}{2}}}_{\text{sample size term}}$$

The p-value then converts this single t-statistic into a probability. It cannot tell you which of the two factors drove t to be large.

**Cohen's d** retains only the first factor. It answers: *how large is the effect in the population, relative to natural variability?*  
**The p-value** conflates both factors. It answers: *how surprising is this data under H₀?* — which can be "very surprising" because d is large **or** because n is large.

#### Practical consequence

| Scenario | p-value says | Cohen's d says | Reality |
|---|---|---|---|
| Large n, tiny effect | Significant (p < 0.05) | Small (d ≈ 0.01) | Statistically significant, practically irrelevant |
| Small n, large effect | Not significant (p > 0.05) | Large (d ≈ 0.8) | Practically important, underpowered study |
| Large n, large effect | Highly significant | Large | Both agree |

This is why reporting only p-values is considered insufficient in modern statistics. Effect sizes (Cohen's d, η², r) and confidence intervals must accompany significance tests to communicate both **whether** an effect exists and **how large** it is.

### ✅ Part (c) Conclusion

| Quantity | What it measures | Depends on n? | Answers |
|---|---|---|---|
| p-value | Rarity of data under H₀ | Yes | Is the result surprising? |
| Cohen's d | Standardised effect magnitude | No | Is the effect large? |
| Power | Probability of detection | Yes | Would we find it again? |

The p-value conflates **effect magnitude** and **sample size** into one number. A large n can produce p < 0.05 even for a trivially small d = 0.01. Cohen's d strips out sample size entirely and reports only the signal-to-noise ratio in the population. In this question both tests share d = 0.25, but the p-values and power differ dramatically — a direct demonstration that p-values should never be read as effect size measures.